### DISCUSSION TITLE
Assembly reference '....dll' was not found or is invalid, times four

### DISCUSSION THREAD

**Discussion by Triazic** [dotnet/fsharp#15577] [Q&A]
Compiler does not work from either inside F# project or running fsc directly. 

Inside F# project:
![image](https://github.com/dotnet/fsharp/assets/38064196/b4e67ac3-7a07-4678-978a-2c7980b17778)

CLI:
![image](https://github.com/dotnet/fsharp/assets/38064196/96cae031-cc54-4ad8-b083-0839fe7fb0ea)

**Repro steps**

Provide the steps required to reproduce the problem:

1. Installed dotnet 7 SDK on windows 10
2. Add package `<PackageReference Include="FSharp.Compiler.Service" Version="43.7.300" />`
3. Write a basic script, SomeScript.fsx:
```
namespace SomeNameSpace

module SomeModule =
  let add (a:float) (b:float):float = a + b
```
4. Attempt to compile the script in a fs project using `FSharp.Compiler.CodeAnalysis`, `FSharpChecker`, `checker.Compile`.  
Fails with errors about the four missing dlls screengrabbed. 
5. Attempt to compile using fsc.dll directly, same four errors. 

**Expected behavior**

It compiles the dll. 

**Actual behavior**

It doesn't compile the dll. 

**Known workarounds**

No idea. 

**Related information**
Hardware, OS etc described in repro steps.

**Comment by vzarytovskii** [comment]
If you need to invoke FSC.dll directly (or use compiler service), it needs all references passed to it in order to work properly, we don't do any discovery inside (only when using msbuild and fsi).

**Comment by baronfel** [comment]
When invoking the compiler explicitly like this, you need to provide all the DLLs to compile _against_. This is normally something that MSBuild does for you as part of dotnet build, but you take that responsibility on yourself when using the APIs.

**Comment by Triazic** [comment]
I will attempt tomorrow. 
Worth pointing out that this is not readily apparent from the docs page people are likely to stumble upon. This example verbatim does not work: https://fsharp.github.io/fsharp-compiler-docs/fcs/compiler.html

Also as an aside, I don't understand why I would need any of those dlls anyway. I'm deffo not doing any winforms or web services in my script. Sigh, dotnet.

**Comment by Triazic** [comment]
Installing .net framework 4.7.2 magically fixed it.
https://dotnet.microsoft.com/en-us/download/dotnet-framework/net472
The same command worked, no need to explicate the allegedly required dlls:
`dotnet "C:\Program Files\dotnet\sdk\7.0.305\FSharp\fsc.dll" -a "SomeScript.fsx" -o "SomeDLL.dll"`

How does this impact deployment? If I want to host the fsc compiler in, say, a game - do I need to distribute .net framework 4.7.2 runtime? Will this approach work on Linux?

Related issue:
https://github.com/dotnet/fsharp/issues/10614#issue-757116229
